kythe/go/util/riegeli/writer.go

```diff
@@ -0,0 +1,202 @@
+	w   io.Writer
+	pos int
+}
+
+// Write implements the io.Writer interface.  The given []byte must be a single
+// non-empty chunk of data.  Interleaving blockHeaders will be written at every
+// 64KiB boundary of the underlying io.Writer.
+func (b *blockWriter) Write(chunk []byte) (n int, err error) {
+	if len(chunk) == 0 {
+		return 0, errors.New("zero-sized chunk")
+	}
+
+	nextBlock := ((b.pos + blockSize - 1) / blockSize) * blockSize
+	remainingBlockSize := nextBlock - b.pos
+
+	// The chunk fits entirely within the current block; write it and return.
+	if remainingBlockSize >= len(chunk) {
+		n, err = b.w.Write(chunk)
+		b.pos += n
+		return
+	}
+
+	blockHeaders := interveningBlockHeaders(b.pos, len(chunk))
+	chunkStart, chunkEnd := b.pos, b.pos+len(chunk)+blockHeaders*blockHeaderSize
+
+	// Fill up the current block with as much of the chunk as possible.
+	if remainingBlockSize > 0 {
+		n, err = b.w.Write(chunk[:remainingBlockSize])
+		b.pos += n
+		if err != nil {
+			return
+		}
+	}
+
+	// For each remaining slice of the chunk of size usableBlockSize, write a
+	// blockHeader and that slice of data.
+	for i, blocksLeft := remainingBlockSize, blockHeaders; blocksLeft > 0; i, blocksLeft = i+usableBlockSize, blocksLeft-1 {
+		// write blockHeader
+		blockStart := b.pos
+		n, err := (&blockHeader{
+			PreviousChunk: uint64(blockStart - chunkStart),
+			NextChunk:     uint64(chunkEnd - blockStart),
+		}).WriteTo(b.w)
+		b.pos += n
+		if err != nil {
+			return b.pos - chunkStart, err
+		}
+
+		l := len(chunk) - i
+		if l > usableBlockSize {
+			// write the maximum size of the chunk possible
+			l = usableBlockSize
+		}
+
+		n, err = b.w.Write(chunk[i : i+l])
+		b.pos += n
+		if err != nil {
+			return b.pos - chunkStart, err
+		}
+	}
+
+	return b.pos - chunkStart, nil
+}
+
+// WriteTo implements the io.WriterTo interface for blockHeaders.
+func (b *blockHeader) WriteTo(w io.Writer) (int, error) {
+	var buf [blockHeaderSize]byte
+	binary.LittleEndian.PutUint64(buf[8:], b.PreviousChunk)
+	binary.LittleEndian.PutUint64(buf[16:], b.NextChunk)
+	binary.LittleEndian.PutUint64(buf[:], hashBytes(buf[8:]))
+	return w.Write(buf[:])
+}
+
+func newRecordChunk(opts *WriterOptions) *recordChunk {
+	return &recordChunk{CompressionType: opts.compressionType()}
+}
+
+// put adds a single record to the Riegeli record chunk.
+func (r *recordChunk) put(rec []byte) error {
+	// TODO(schroederc): compression
+	size := uint64(len(rec))
+
+	var buf [binary.MaxVarintLen64]byte
+	n := binary.PutUvarint(buf[:], size)
+
+	r.CompressedSizes = append(r.CompressedSizes, buf[:n]...)
+	r.CompressedValues = append(r.CompressedValues, rec...)
+
+	r.decodedSize += size
+	r.numRecords++
+	return nil
+}
+
+// encode returns the binary-encoding of the Riegeli record chunk.
+func (r *recordChunk) encode() []byte {
+	// TODO(schroederc): compression
+	buf := bytes.NewBuffer(make([]byte, 0, 1+binary.MaxVarintLen64+len(r.CompressedSizes)+len(r.CompressedValues)))
```

Comment:
Again maybe for the future: Consider passing in a buffer here to avoid reallocating (the writer doesn't support concurrent writers anyway)

Reply:
✓